git-changebar/src/gcb-plugin.c

```diff
@@ -1086,6 +1106,131 @@ on_kb_goto_next_hunk (guint kb)
   }
 }
 
+static void
+insert_buf_range (GeanyDocument *doc,
+                  const git_buf *old_contents,
+                  gint           pos,
+                  gint           old_start,
+                  gint           old_lines)
+{
+  ScintillaObject *old_sci     = editor_create_widget (doc->editor);
+  gchar           *old_buf     = old_contents->ptr;
+  gsize            old_buf_len = old_contents->size;
+  gboolean         free_buf    = FALSE;
+  gint             old_pos_start, old_pos_end;
+  gchar           *old_range;
+  
+  /* convert the buffer to UTF-8 if necessary */
+  if (encoding_needs_coversion (doc->encoding)) {
+    free_buf = convert_encoding_inplace (&old_buf, &old_buf_len, free_buf,
+                                         "UTF-8", doc->encoding, NULL);
+  }
+  
+  scintilla_send_message (old_sci, SCI_ADDTEXT, old_buf_len, (glong) old_buf);
+
+  old_pos_start = sci_get_position_from_line (old_sci, old_start);
+  old_pos_end = sci_get_position_from_line (old_sci, old_start + old_lines);
+  old_range = sci_get_contents_range (old_sci, old_pos_start, old_pos_end);
+
+  sci_insert_text(doc->editor->sci, pos, old_range);
+
+  g_free(old_range);
+  
+  if (free_buf) {
+    g_free (old_buf);
+  }
+
+  g_object_ref_sink (old_sci);
+  g_object_unref (old_sci);
+}
+
+static int
+undo_hunk_diff_hunk_cb (const git_diff_delta *delta,
+                        const git_diff_hunk  *hunk,
+                        void                 *udata)
+{
+  UndoHunkData *data = udata;
+  gboolean first_line_removed = is_first_line_removed (data->line, hunk->new_start, hunk->new_lines);
+
+  if (first_line_removed ||
+      (data->line >= hunk->new_start &&
+       data->line < hunk->new_start + MAX (1, hunk->new_lines))) {
+    data->old_start = hunk->old_start;
+    data->old_lines = hunk->old_lines;
+    data->new_start = hunk->new_start;
+    data->new_lines = hunk->new_lines;
+    data->found = TRUE;
+    data->first_line_removed = first_line_removed;
+    return 1;
+  }
+  
+  return 0;
+}
+
+static void
+undo_hunk_cb (const gchar *path,
+              git_buf     *contents,
+              gpointer     udata)
+{
+  UndoHunkData  *data = udata;
+  GeanyDocument *doc  = document_get_current ();
+  
+  if (doc && doc->id == data->doc_id && contents) {
+    diff_buf_to_doc (contents, doc, undo_hunk_diff_hunk_cb, data);
+
+    if (data->found) {
+      ScintillaObject *sci = doc->editor->sci;
+
+      sci_start_undo_action(sci);
+
+      if (data->new_lines > 0) {
+        gint pos = sci_get_position_from_line(sci, data->new_start - 1);
+        sci_set_target_start(sci, pos);
+        pos = sci_get_position_from_line(sci, data->new_start + data->new_lines - 1);
+        sci_set_target_end(sci, pos);
+        sci_replace_target(sci, "", FALSE);
+      }
+
+      if (data->old_lines > 0) {
+        gint line = sci_get_current_line (sci);
```

Comment:
> `gint pos = sci_get_position_from_line (sci, data->new_start - 1);` returns the pos on the previous line which is wrong when undoing deleted-only diff - I need to advance it to the next line if `new_lines == 0` (see a few lines below).

Indeed, I was confused at first.  I suggest 00e37457c20c94e23b1eef8ea98ea1977c446635

> Just to understand - what does scintilla return from sci_get_current_line() when you right-click somewhere - I'd expect it doesn't change the current line, does it?

No it doesn't, but Geany gives us the position of the right click, and other features in the menu work at the click position, not the caret position.  So for the feature I want to do the same: propose undoing the diff at the right click position, not at the caret one.  See 0c4415dbe61792de805e45e72433e4feaac988f5 for a proposal.